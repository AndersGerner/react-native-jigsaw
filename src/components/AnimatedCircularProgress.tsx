import * as React from "react";
import { Animated, Easing, EasingFunction } from "react-native";
import CircularProgress, {
  Props as CircularProgressProps,
} from "./CircularProgress";
const AnimatedProgress = Animated.createAnimatedComponent(CircularProgress);

interface Props extends CircularProgressProps {
  duration?: number;
  easing?: EasingFunction;
  prefill?: number;
  useNativeDriver?: boolean;
  tintColorSecondary?: string;
  onAnimationComplete?: Animated.EndCallback | undefined;
}

const AnimatedCircularProgress: React.FC<Props> = ({
  duration = 500,
  easing = Easing.out(Easing.ease),
  fill,
  prefill = 0,
  useNativeDriver = false,
  tintColorSecondary,
  onAnimationComplete,
  tintColor = "black",
  ...other
}) => {
  const [fillAnimation] = React.useState<Animated.Value>(
    new Animated.Value(prefill)
  );

  //   const reAnimate = (prefill, toVal, dur, ease) => {
  //     setFillAnimation(new Animated.Value(prefill));
  //     animate(toVal, dur, ease);
  //   };

  const animate = React.useCallback(
    (
      toVal: number = -1,
      dur?: number,
      ease?: EasingFunction
    ): Animated.CompositeAnimation => {
      const toValue = toVal >= 0 ? toVal : fill;
      const dura = dur || duration;
      const eas = ease || easing;
      const useNative = useNativeDriver;

      const anim = Animated.timing(fillAnimation, {
        useNativeDriver: useNative,
        toValue,
        easing: eas,
        duration: dura,
      });
      anim.start(onAnimationComplete);

      return anim;
    },
    [
      duration,
      easing,
      fill,
      useNativeDriver,
      fillAnimation,
      onAnimationComplete,
    ]
  );

  const animateColor = () => {
    if (!tintColorSecondary) {
      return tintColor;
    }

    const tintAnimation = fillAnimation.interpolate({
      inputRange: [0, 100],
      outputRange: [tintColor, tintColorSecondary],
    });

    return tintAnimation;
  };

  React.useEffect(() => {
    animate();
  }, [fill, animate]);

  return (
    <AnimatedProgress
      {...other}
      fill={fillAnimation}
      tintColor={animateColor()}
    />
  );
};

export default AnimatedCircularProgress;
