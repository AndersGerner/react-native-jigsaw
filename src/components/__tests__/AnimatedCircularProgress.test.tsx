import * as React from "react";
import renderer from "react-test-renderer";
import AnimatedCircularProgress from "../AnimatedCircularProgress";

it("renders animated circular progress", () => {
  const tree = renderer
    .create(<AnimatedCircularProgress fill={50} size={30} width={200} />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
