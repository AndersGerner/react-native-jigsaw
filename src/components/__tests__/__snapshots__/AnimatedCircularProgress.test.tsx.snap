// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders animated circular progress 1`] = `
<View>
  <RNSVGSvgView
    bbHeight={30}
    bbWidth={30}
    focusable={false}
    height={30}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        Object {
          "flex": 0,
          "height": 30,
          "width": 30,
        },
      ]
    }
    width={30}
  >
    <RNSVGGroup>
      <RNSVGGroup
        matrix={
          Array [
            6.123233995736766e-17,
            1,
            -1,
            6.123233995736766e-17,
            30,
            -1.7763568394002505e-15,
          ]
        }
      />
    </RNSVGGroup>
  </RNSVGSvgView>
</View>
`;
